src/error_handling.c

```diff
@@ -0,0 +1,259 @@
+  utilities_error_state_t * __utilities_error_state =
+    (utilities_error_state_t *)pthread_getspecific(__utilities_error_state_key);
+  char * __utilities_error_string = (char *)pthread_getspecific(__utilities_error_string_key);
+#endif
+  if (!__utilities_error_is_set(__utilities_error_state)) {
+    return;
+  }
+  size_t bytes_it_would_have_written = snprintf(
+    NULL, 0,
+    __error_format_string,
+    __utilities_error_state->message,
+    __utilities_error_state->file,
+    __utilities_error_state->line_number);
+  utilities_allocator_t allocator = __utilities_error_state->allocator;
+  __utilities_error_string =
+    (char *)allocator.allocate(bytes_it_would_have_written + 1, allocator.state);
+#ifdef UTILITIES_THREAD_LOCAL_PTHREAD
+  pthread_setspecific(__utilities_error_string_key, __utilities_error_string);
+#endif
+  if (!__utilities_error_string) {
+#if UTILITIES_REPORT_ERROR_HANDLING_ERRORS
+    // utilities_allocate failed, but fwrite might work?
+    SAFE_FWRITE_TO_STDERR(
+      "[rmw|error_handling.c:" UTILITIES_STRINGIFY(__LINE__)
+      "] failed to allocate memory for the error string\n");
+#endif
+    return;
+  }
+  snprintf(
+    __utilities_error_string, bytes_it_would_have_written + 1,
+    __error_format_string,
+    __utilities_error_state->message,
+    __utilities_error_state->file,
+    __utilities_error_state->line_number);
+  // The Windows version of snprintf does not null terminate automatically in all cases.
+  __utilities_error_string[bytes_it_would_have_written] = '\0';
+}
+
+const char *
+utilities_get_error_string()
+{
+#ifdef UTILITIES_THREAD_LOCAL_PTHREAD
+  char * __utilities_error_string = (char *)pthread_getspecific(__utilities_error_string_key);
+#endif
+  if (!__utilities_error_string) {
+    format_error_string();
+  }
+  return __utilities_error_string;
+}
+
+bool
+__utilities_error_is_set(utilities_error_state_t * error_state)
+{
+  return error_state != NULL;
+}
+
+bool
+utilities_error_is_set()
+{
+#ifdef UTILITIES_THREAD_LOCAL_PTHREAD
+  utilities_error_state_t * __utilities_error_state =
+    (utilities_error_state_t *)pthread_getspecific(__utilities_error_state_key);
+#endif
+  return __utilities_error_is_set(__utilities_error_state);
+}
+
+const char *
+utilities_get_error_string_safe()
+{
+  if (!utilities_error_is_set()) {
+    return "error not set";
+  }
+  return utilities_get_error_string();
+}
+
+void
+__utilities_reset_error_string(char ** error_string_ptr, utilities_allocator_t allocator)
+{
+  char * error_string = *error_string_ptr;
+  if (error_string) {
+    if (error_string) {
+      allocator.deallocate(error_string, allocator.state);
+    }
+  }
+  *error_string_ptr = NULL;
+}
+
+void
+__utilities_reset_error(utilities_error_state_t ** error_state_ptr_ptr)
+{
+  if (error_state_ptr_ptr) {
+    utilities_error_state_t * error_state_ptr = *error_state_ptr_ptr;
+    if (error_state_ptr) {
+      utilities_allocator_t allocator = error_state_ptr->allocator;
```

Comment:
`error_state_ptr` may be null, but `error_state_ptr->allocator` cannot be null. `error_state_ptr->allocator->deallocate` can be null though. The assumption is (and I think it holds) that if `error_state_ptr` is not null, then the allocator will be valid as well. I can make it more defensive though.